### PR TITLE
Introduce `CharSequenceRules` Refaster rule collection

### PR DESCRIPTION
**Pull request by rickie**
This aims to solve the following issue: https://github.com/PicnicSupermarket/error-prone-support/issues/1394. 

@timtebeek mentioned that OpenRewrite has some checks that ensure that only the correct recipes are loaded for the Java version it's running on. For us I also don't see any problems for running on our code. The only advantage is that we will be matching more now :).

### CODE REVIEW

File: error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/CharSequenceRulesTestInput.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package tech.picnic.errorprone.refasterrules;
+
+import com.google.common.collect.ImmutableSet;
+import tech.picnic.errorprone.refaster.test.RefasterRuleCollectionTestCase;
+
+final class CharSequenceRulesTest implements RefasterRuleCollectionTestCase {
+  ImmutableSet<Boolean> testCharSequenceIsEmpty() {
+    return ImmutableSet.of(
+        ((CharSequence) "foo").length() == 0,
```
Comment (rickie) [inline error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/CharSequenceRulesTestInput.java:9]: Maybe there is a better way, but now we know for sure its a `CharSequence` :D

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/CharSequenceRules.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+final class CharSequenceRules {
+  private CharSequenceRules() {}
+
+  /**
+   * Prefer {@link CharSequence#isEmpty()} over alternatives that consult the char sequence's
+   * length.
+   */
+  static final class CharSequenceIsEmpty {
+    @BeforeTemplate
+    boolean before(CharSequence ch) {
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/CharSequenceRules.java]:
```suggestion
    boolean before(CharSequence charSequence) {	
```
As `ch` generally denotes "char"`. (Unifying the variable naming approach is also on my list, somewhere ;).)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/StringRules.java [hunk 3]
```diff
@@ -32,6 +32,7 @@ private StringRules() {}
   // XXX: Now that we build with JDK 15+, this rule can be generalized to cover all `CharSequence`
   // subtypes. This does require a mechanism (perhaps an annotation, or a separate Maven module) to
   // make sure that non-String expressions are rewritten only if client code also targets JDK 15+.
+  @SuppressWarnings("CharSequenceIsEmpty" /* This is a more specific template. */)
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/StringRules.java]: The scope of this suppression should be restricted to the `@BeforeTemplate`.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/StringRules.java [hunk 4]
```diff
@@ -32,6 +32,7 @@ private StringRules() {}
   // XXX: Now that we build with JDK 15+, this rule can be generalized to cover all `CharSequence`
   // subtypes. This does require a mechanism (perhaps an annotation, or a separate Maven module) to
   // make sure that non-String expressions are rewritten only if client code also targets JDK 15+.
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/StringRules.java]: This comment should be updated.

File: error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/CharSequenceRulesTestInput.java (hunk 1 above)
Comment (Stephan202) [reply error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/CharSequenceRulesTestInput.java]: `new StringBuilder("foo")`, `str.subSequence(0, 1)`, ... :)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/CharSequenceRules.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+/** Refaster rules related to expressions dealing with {@link CharSequence}s. */
+@OnlineDocumentation
+final class CharSequenceRules {
+  private CharSequenceRules() {}
+
+  /**
+   * Prefer {@link CharSequence#isEmpty()} over alternatives that consult the char sequence's
+   * length.
+   */
+  static final class CharSequenceIsEmpty {
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/CharSequenceRules.java:20]: We should document this rule's relation to `StringIsEmpty`.

File: error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/CharSequenceRulesTestInput.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        ((CharSequence) "bar").length() <= 0,
+        ((CharSequence) "baz").length() < 1,
+        ((CharSequence) "foo").length() != 0,
+        ((CharSequence) "bar").length() > 0,
+        ((CharSequence) "baz").length() >= 1);
```
Comment (Stephan202) [inline error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/CharSequenceRulesTestInput.java]: `"qux"`, `"quux"`, `"corge"` (we inconsistently sometimes also include `"quuz"`; something to unify another day ;).)

**Review by Stephan202 — APPROVED** [review]
Rebased and added a commit. Suggested commit message:
```
Introduce `CharSequenceRules` Refaster rule collection (#1480)

Resolves #1394.
```

**Review by mohamedsamehsalah — APPROVED** [review]
🚀

**Pull request merged**